collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py

```diff
@@ -1,58 +1,99 @@
+        if kv.startswith('logits'):
             for v_tensor in v:
                 for logit_tensor in v_tensor:
-                    logits_lists.append(logit_tensor.detach().cpu().tolist())
+                    all_logits.append(tensor2list(logit_tensor))
+
+        elif kv.startswith('labels'):
+            for v_tensor in v:
+                for label_tensor in v_tensor:
+                    all_labels.extend(tensor2list(label_tensor))
 
-        if 'seq_ids' in kv:
+        elif kv.startswith('subtokens_mask'):
             for v_tensor in v:
-                for seq_id_tensor in v_tensor:
-                    seq_ids.append(seq_id_tensor.detach().cpu().tolist())
+                for subtokens_mask_tensor in v_tensor:
+                    all_subtokens_mask.extend(
+                        tensor2list(subtokens_mask_tensor))
 
-    correct_labels, incorrect_labels, correct_preds, total_preds, \
-        total_correct = eval_data_layer.eval_preds(logits_lists, seq_ids)
+    all_preds = list(np.argmax(np.asarray(all_logits), 2).flatten())
+    global_vars["all_preds"].extend(all_preds)
+    global_vars["all_labels"].extend(all_labels)
+    global_vars["all_subtokens_mask"].extend(all_subtokens_mask)
 
-    global_vars["correct_labels"] += correct_labels
-    global_vars["incorrect_labels"] += incorrect_labels
-    global_vars["correct_preds"] += correct_preds
-    global_vars["total_preds"] += total_preds
-    global_vars["total_correct"] += total_correct
 
+def eval_epochs_done_callback(global_vars,
+                              label_ids,
+                              graph_fold=None,
+                              none_label_id=0):
+    labels = np.asarray(global_vars['all_labels'])
+    preds = np.asarray(global_vars['all_preds'])
+    subtokens_mask = np.asarray(global_vars['all_subtokens_mask'])
 
-def eval_epochs_done_callback(global_vars):
+    labels = labels[subtokens_mask]
+    preds = preds[subtokens_mask]
 
-    correct_labels = global_vars["correct_labels"]
-    incorrect_labels = global_vars["incorrect_labels"]
-    correct_preds = global_vars["correct_preds"]
-    total_preds = global_vars["total_preds"]
-    total_correct = global_vars["total_correct"]
+    accuracy = sum(labels == preds) / labels.shape[0]
+    logger.info(f'Accuracy: {accuracy}')
 
-    accuracy = correct_labels / (correct_labels + incorrect_labels)
+    # print predictions and labels for a small random subset of data
+    sample_size = 20
+    i = 0
+    if preds.shape[0] > sample_size + 1:
+        i = random.randint(0, preds.shape[0] - sample_size - 1)
+    logger.info("Sampled preds: [%s]" % list2str(preds[i:i+sample_size]))
+    logger.info("Sampled labels: [%s]" % list2str(labels[i:i+sample_size]))
+    logger.info(classification_report(labels, preds, target_names=label_ids))
 
-    p = correct_preds / total_preds
-    r = correct_preds / total_correct
-    f1 = 2 * p * r / (p + r)
+    # calculate and plot confusion_matrix
+    if graph_fold:
```

Comment:
i think we should have an option in the argument to choose whether to show confusion_matrix. technically you can still show confusion matrix even if you don't have graph_fold. You only need graph_fold if you want to save the image.